demos/burgers-goal_oriented.py

```diff
@@ -0,0 +1,425 @@
+
+    return solver
+
+
+def get_qoi(mesh_seq, i):
+    R = FunctionSpace(mesh_seq[i], "R", 0)
+    dt = Function(R).assign(mesh_seq.time_partition.timesteps[i])
+
+    def time_integrated_qoi(t):
+        u = mesh_seq.fields["u"][0]
+        return dt * inner(u, u) * ds(2)
+
+    return time_integrated_qoi
+
+
+# We use the same mesh setup as in `the previous demo <./burgers2.py.html>`__ and the
+# same time partitioning, except that we export every timestep rather than every other
+# timestep. ::
+
+n = 32
+meshes = [UnitSquareMesh(n, n, diagonal="left"), UnitSquareMesh(n, n, diagonal="left")]
+end_time = 0.5
+dt = 1 / n
+
+num_subintervals = len(meshes)
+time_partition = TimePartition(
+    end_time,
+    num_subintervals,
+    dt,
+    fields,
+    num_timesteps_per_export=1,
+)
+
+mesh_seq = GoalOrientedMeshSeq(
+    time_partition,
+    meshes,
+    get_function_spaces=get_function_spaces,
+    get_initial_condition=get_initial_condition,
+    get_form=get_form,
+    get_solver=get_solver,
+    get_qoi=get_qoi,
+    qoi_type="time_integrated",
+)
+
+# Compared to the previous adaptation demo, this adaptor depends on adjoint solution data
+# as well as forward solution data. For simplicity, we begin by using Goalie's inbuilt
+# isotropic metric function. ::
+
+
+def adaptor(mesh_seq, solutions=None, indicators=None):
+    metrics = []
+    complexities = []
+
+    indicators = mesh_seq.indicators
+
+    # Ramp the target average metric complexity per timestep
+    base, target, iteration = 400, 1000, mesh_seq.fp_iteration
+    mp = {
+        "dm_plex_metric": {
+            "target_complexity": ramp_complexity(base, target, iteration),
+            "p": 1.0,
+            "h_min": 1.0e-04,
+            "h_max": 1.0,
+        }
+    }
+
+    # Construct the metric on each subinterval
+    for i, mesh in enumerate(mesh_seq):
+        dt = mesh_seq.time_partition.timesteps[i]
+
+        # Define the Riemannian metric
+        P1_ten = TensorFunctionSpace(mesh, "CG", 1)
+
+        # At each timestep, recover metric of the solution
+        # vector. Then time integrate over the contributions
+        metrics_i = []
+
+        # Calculate metric at each timestep
+        for indi in indicators["u"][i]:
+            # get local indicator
+            # indi = indicators["u"][i][j]
+            # local instance of Riemanian metric
+            metric_j = RiemannianMetric(P1_ten)
+
+            # reset parameters
+            metric_j.set_parameters(mp)
+
+            # Deduce an isotropic metric from the error indicator field
+            metric_j.compute_isotropic_metric(error_indicator=indi, interpolant="L2")
+            metric_j.normalise()
+
+            # append the metric for the step in the time partition
+            metrics_i.append(metric_j)
+
+        # set the first metric to the base
+        metric = metrics_i[0]
+        # all the other metrics
+        metric.average(*metrics_i[1:], weights=[dt] * len(metrics_i))
```

Comment:
I think this might be confusing to new users, but it's not your fault - I think it would make sense to add a separate `average` and `intersect` function to avoid this strange notation. If you agree, maybe we could open a new issue for this?

Reply:
@ddundo I think there is already an issue opened on this? See issue: https://github.com/mesh-adaptation/animate/issues/84.

Reply:
Ah sorry! I didn't read Joe's previous review :)